### PR TITLE
Create a DIDMethod interface and Implement it

### PR DESCRIPTION
**Pull request by andresuribe87**
# Overview
This PR unifies the two implementations we had of DIDKey, and declares an interface to be implemented by any DIDMethod.

# Description
This is a step towards reducing confusion in our current codebase. It also adds tests and moves tests around for the DIDKey pieces. A resolver for DIDKey is implemented as well. 

Note that because of the ambiguous codecs in multicodecs, this code currently only works for keys generated from the Ed25519 curve.

# How Has This Been Tested?
Unit tests

### CODE REVIEW

File: credentials/src/main/kotlin/SSI.kt [hunk 1]
```diff
@@ -182,6 +115,17 @@ public class VerifiableCredential private constructor() {
...
 }
 
+private fun JWSAlgorithm.toVerifier(publicKeyJWK: JWK): JWSVerifier? {
+  return when (this) {
+    JWSAlgorithm.EdDSA -> Ed25519Verifier(publicKeyJWK.toOctetKeyPair())
+    JWSAlgorithm.ES256K -> ECDSAVerifier(publicKeyJWK.toECKey())
+    JWSAlgorithm.ES256 -> ECDSAVerifier(publicKeyJWK.toECKey())
+    JWSAlgorithm.ES384 -> ECDSAVerifier(publicKeyJWK.toECKey())
+    JWSAlgorithm.ES512 -> ECDSAVerifier(publicKeyJWK.toECKey())
+    else -> null
```
Comment (nitro-neal) [inline credentials/src/main/kotlin/SSI.kt:125]: should we throw an exception ?

File: dids/src/main/kotlin/DIDKeyMethod.kt [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    require(opts is CreateDIDKeyOptions)
+    require(!opts.publicJwk.isPrivate)
+    return DIDKeyCreator(opts.publicJwk)
+  }
+}
+
+public class DIDKeyCreationMetadata(public val jwk: JWK) : DIDCreationMetadata
+
+private class DIDKeyCreator(
+  private val jwk: JWK,
```
Comment (nitro-neal) [inline dids/src/main/kotlin/DIDKeyMethod.kt:82]: something is up with this formatting?

File: dids/src/main/kotlin/DIDKeyMethod.kt [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    private fun createSignatureMethod(did: DID): VerificationMethod {
+      val multibaseValue = did.methodSpecificId
+      val decodedMultibase = Multibase.decode(multibaseValue)
+
+      // TODO: The decode function is not unambiguous. See https://github.com/richardbergquist/java-multicodec#current-workarounds
+      val decodedData = MulticodecEncoder.decode(decodedMultibase)
+      val multicodecValue = decodedData.codec
+      val rawPublicKeyBytes = decodedData.dataAsBytes
+
+      require(rawPublicKeyBytes.size == expectedPublicKeySize(multicodecValue))
```
Comment (nitro-neal) [inline dids/src/main/kotlin/DIDKeyMethod.kt:109]: nice

File: dids/src/test/kotlin/DIDKeyMethodTest.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import web5.dids.CreateDIDKeyOptions
+import web5.dids.DIDKeyMethod
+import web5.dids.DIDKeyResolver
+import kotlin.test.assertContains
+
+class DIDKeyMethodTest {
+
+  @Test
+  fun authorizeReturnsTrue() {
+    val privateJWK = OctetKeyPairGenerator(Curve.Ed25519).generate()
```
Comment (nitro-neal) [inline dids/src/test/kotlin/DIDKeyMethodTest.kt:14]: Maybe make a for loop for tests with the other algs and a non supported alg

**Review by nitro-neal — APPROVED** [review]
looks good to me

**Comment by mistermoe** [comment]
closing as a result of meging #34

**Pull request closed without merging**